text/2018-11-12-backup-restore-replication-with-raft-learner.md

```diff
@@ -0,0 +1,168 @@
+
+For the Backup, we only need to retrieve the metadata, it is lightweight, so we can do the backup frequently. 
+
+## Restore
+
+After Backup, we can get a snapshot of the whole cluster, we can use it for restore. 
+
+For a region, if we get a snapshot file and its following Raft logs, we can first apply the snapshot, and then apply the Raft logs in order, and the region will have the consistent data eventually. We name this flow Replay later.
+
+But here we must consider the Replay order of different regions. For example, we have region 1 and 2, and 2 is created through a Spit command in the logs of region 1 at index 100. 
+
+The origin range of region 1 is [a, c), after Split, the range is [b, c) and region 2’s is [a, b). 
+
+```
+                                            checkpoint 
+                                                |      
++---------+      +---------+---+---+---+---+---+|      
+|Region 1 |      |Snapshot |11 |...|100|101 102||      
++---------+      |Index 10 +---+---+---+---+---+|      
+                 +---------+         |          |      
+                              Split  |          |      
+                              1 -> 2 + 1        |      
+                                     |          |      
+                                     |          |      
++---------+                          | +---+---+|      
+|Region 2 |                          +>|1  |2  ||      
++---------+                            +---+---+|      
+                                                v      
+```                                                                             
+
+When we do the checkpoint, the checkpoint has already contained region 1 and region 2. and region 1 will be replayed from the Snapshot, region 2 will be replayed from the first log. But here is a problem when we replay region 2 before region 1. 
+
+Assume the command in the log 1 of region 2 is `SET b = 10`, but the command in the log 11 of region 1 is `SET b = 9`. So if region 2 is replayed first, the `b` will have a wrong value. To solve this problem, we need to build a Replay order, the implementation may be easy - just to record the event.
+
+For Split 1 -> 2 + 1, region 2 must record a event that `region 2 is split from region 1 at index 100`. When region 2 wants to replay, it finds that - “oh, region 1’s hasn’t applied to 100 yet, so let’s wait”.
+
+For merge 1 + 2 -> 1, region 1 must also persist the similar event, and it must wait for region 2 replaying then start to replay later.
+
+Using this way has another benefit, now we must sync the state machine forcibly when we meet Split, Merge, or other Admin commands because we have no way to recover the cluster with only replaying Raft logs. Sync can cause latency increased sometimes, so removing this can help us increase the performance.
```

Comment:
```suggestion
Using this way has another benefit, now we must sync the state machine forcibly when we meet Split, Merge, or other Admin commands because we have no way to recover the cluster with only replaying Raft logs. Sync can cause latency increase sometimes, so removing this can help us increase the performance.
```